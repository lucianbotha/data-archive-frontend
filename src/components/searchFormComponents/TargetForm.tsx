import * as React from "react";
import targetPosition from "target-position";
import { ITarget } from "../../utils/ObservationQueryParameters";
import {
  isFloat,
  validateDec,
  validateName,
  validateRa
} from "../../utils/validators";
import { InnerMainGrid, MainGrid, SubGrid } from "../basicComponents/Grids";
import InputField from "../basicComponents/InputField";
import SelectField from "../basicComponents/SelectField";

class TargetForm extends React.Component<
  { target: ITarget; onChange: any },
  any
> {
  resolve = () => {
    const { target, onChange } = this.props;
    const resolver = target.resolver || "Simbad";

    targetPosition(target.name || "", [resolver])
      .then(p => {
        if (p) {
          onChange({
            ...target,
            declination: `${p.declination}`,
            rightAscension: `${p.rightAscension}`
          });
        } else {
          onChange({
            ...target,
            errors: {
              ...target.errors,
              name: `The target name ${target.name} could not be resolved.`
            }
          });
        }
      })
      .catch(err => {
        onChange({
          ...target,
          errors: {
            ...target.errors,
            name: err.message
          }
        });
      });
  };
  render() {
    const { target, onChange } = this.props;
    const targetChange = (
      e: React.FormEvent<HTMLSelectElement | HTMLInputElement>
    ) => {
      const name = e.currentTarget.name;
      const value = e.currentTarget.value;
      onChange({
        ...target,
        [name]: value,
        errors: {
          ...target.errors,
          [name]: ""
        }
      });
    };
    return (
      <>
        <MainGrid>
          <SubGrid>
            <p>Target name</p>
            <InputField
              name={"name"}
              value={target.name || ""}
              error={target.errors.name || ""}
              onChange={targetChange}
            />
          </SubGrid>
          <SubGrid>
            <InnerMainGrid>
              <SubGrid>
                <p>Resolver</p>
                <SelectField
                  options={["Simbad", "NED", "VizieR"]}
                  name={"resolver"}
                  value={target.resolver || "Simbad"}
                  onChange={targetChange}
                />
              </SubGrid>
              <SubGrid>
                <br />
                <input
                  className="button is-info"
                  type="button"
                  value="resolve"
                  onClick={this.resolve}
                />
              </SubGrid>
            </InnerMainGrid>
          </SubGrid>
        </MainGrid>

        <MainGrid>
          <SubGrid>
            <p>Right ascension</p>
            <InputField
              name={"rightAscension"}
              value={target.rightAscension || ""}
              onChange={targetChange}
              error={target.errors.rightAscension}
            />
          </SubGrid>
          <SubGrid>
            <p>Declination</p>
            <InputField
              name={"declination"}
              value={target.declination || ""}
              onChange={targetChange}
              error={target.errors.declination}
            />
          </SubGrid>
        </MainGrid>
        <MainGrid>
          <SubGrid>
            <p>Search radius</p>
            <InputField
              name={"searchConeRadius"}
              value={target.searchConeRadius || ""}
              onChange={targetChange}
              error={target.errors.searchConeRadius}
            />
          </SubGrid>
          <SubGrid>
            <p>Radius units</p>
            <SelectField
              options={["Arcseconds", "Arcminutes", "Degrees"]}
              name={"radiusUnits"}
              onChange={targetChange}
              value={target.searchConeRadiusUnits || "Arcseconds"}
            />
          </SubGrid>
        </MainGrid>
      </>
    );
  }
}

/**
 */
export const validatedTarget = async (target: ITarget) => {
  let raDecChangeError;
  if (target.name && target.name !== "") {
    raDecChangeError = await targetPosition(target.name, [
      target.resolver || "Simbad"
    ])
      .then(p => {
        if (p) {
          return {
            declination:
              `${p.declination}` !== target.declination
                ? `The declination differs from that returned by ${
                    target.resolver
                  } for ${target.name}.`
                : "",
            rightAscension:
              `${p.rightAscension}` !== target.rightAscension
                ? `The right ascension differs from that returned by ${
                    target.resolver
                  } for ${target.name}.`
                : ""
          };
        }
      })
      .catch();
  }
  return {
    ...target,
    errors: {
      declination:
        raDecChangeError && raDecChangeError.declination
          ? raDecChangeError.declination
          : validateDec(target.declination || ""),
      name: validateName(target.name || ""),
      rightAscension:
        raDecChangeError && raDecChangeError.rightAscension
          ? raDecChangeError.rightAscension
          : validateRa(target.rightAscension || ""),
      searchConeRadius: isFloat(target.searchConeRadius || "")
        ? ""
        : "The search radius must be a positive number."
    }
  };
};
export default TargetForm;
