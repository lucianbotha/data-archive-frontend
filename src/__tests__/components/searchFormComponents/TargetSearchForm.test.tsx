import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import TargetForm from "../../../components/searchFormComponents/TargetForm";

const onChange = jest.fn();

describe("TargetForm", () => {
  const wrapper = mount(
    <TargetForm target={{ errors: {} }} onChange={onChange} />
  );

  it("should render", () => {
    expect(wrapper).toBeDefined();
  });

  it("should contains 4 inputs and 2 selects and a button", () => {
    const inputs = wrapper.find("input.input");
    expect(inputs.length).toEqual(4);
    const select = wrapper.find("select");
    expect(select.length).toEqual(2);
    const button = wrapper.find("button.button");
    expect(button.length).toEqual(1);
  });

  it("should contains input of name (name, resolver, rightAscension, declination, searchConeRadius and radiusUnits", () => {
    const name = wrapper.find("input").get(0).props.name;
    expect(name).toEqual("name");

    const resolver = wrapper.find("select").get(0).props.name;
    expect(resolver).toEqual("resolver");

    const rightAscension = wrapper.find("input").get(1).props.name;
    expect(rightAscension).toEqual("rightAscension");

    const declination = wrapper.find("input").get(2).props.name;
    expect(declination).toEqual("declination");

    const searchConeRadius = wrapper.find("input").get(3).props.name;
    expect(searchConeRadius).toEqual("searchConeRadius");

    const radiusUnits = wrapper.find("select").get(1).props.name;
    expect(radiusUnits).toEqual("radiusUnits");
  });

  it("render correctly", () => {
    expect(
      toJson(
        shallow(<TargetForm target={{ errors: {} }} onChange={onChange} />)
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <TargetForm
            target={{
              errors: {},
              name: "hello"
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
    expect(
      toJson(
        shallow(
          <TargetForm
            target={{
              errors: {
                name: "this is an error"
              },
              searchConeRadius: "12.1"
            }}
            onChange={onChange}
          />
        )
      )
    ).toMatchSnapshot();
  });
});

describe("TargetForm onChange", () => {
  const wrapper = mount(
    <TargetForm target={{ errors: {} }} onChange={onChange} />
  );
  it("should call onChange with targetName", () => {
    const target = wrapper.find("input.target-name-input");
    target.simulate("change", "");
    expect(onChange).toBeCalled();
  });
  it("should call onChange with right ascension", () => {
    let value: any;
    let event: any;
    const resolverForm = wrapper.find("input.right-ascension-input");
    const resolver = resolverForm.find("input");
    value = "123.1";
    event = { target: { value, name: "rightAscension" } };
    resolver.simulate("change", event);
    expect(onChange).toBeCalledWith({
      errors: { rightAscension: "" },
      rightAscension: "123.1"
    });

    value = "";
    event = { target: { value, name: "rightAscension" } };
    resolver.simulate("change", event);
    expect(onChange).toBeCalledWith({
      errors: { rightAscension: "" },
      rightAscension: ""
    });

    value = undefined;
    event = { target: { value, name: "rightAscension" } };
    resolver.simulate("change", event);
    expect(onChange).toBeCalledWith({
      errors: { rightAscension: "" },
      rightAscension: ""
    });
  });

  it("should call onChange for selectors as well", () => {
    const resolver = wrapper.find("div.resolver-select");
    resolver.simulate("change", "");
    expect(onChange).toBeCalled();

    const rUnits = wrapper.find("div.radius-units-select");
    rUnits.simulate("change", "");
    expect(onChange).toBeCalled();
  });
});
