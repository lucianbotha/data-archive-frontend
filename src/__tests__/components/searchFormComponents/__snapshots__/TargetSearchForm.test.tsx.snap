// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TargetForm render correctly 1`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Target name
      </p>
      <InputField
        className="target-name-input"
        disabled={false}
        error=""
        name="name"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div>
          <p>
            Resolver
          </p>
          <SelectField
            className="resolver-select"
            name="resolver"
            onChange={[Function]}
            value="Simbad"
          >
            <option
              value="Simbad"
            >
              Simbad
            </option>
            <option
              value="NED"
            >
              NED
            </option>
            <option
              value="VizieR"
            >
              VizieR
            </option>
          </SelectField>
        </styled.div>
        <styled.div>
          <br />
          <button
            className="button is-info false"
            disabled={false}
            onClick={[Function]}
            title=""
            type="button"
          >
            resolve
          </button>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Right ascension
      </p>
      <InputField
        className="right-ascension-input"
        disabled={false}
        name="rightAscension"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Declination
      </p>
      <InputField
        className="declination-input"
        disabled={false}
        name="declination"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Search radius
      </p>
      <InputField
        className="search-cone-radius-input"
        name="searchConeRadius"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Radius units
      </p>
      <SelectField
        className="radius-units-select"
        name="radiusUnits"
        onChange={[Function]}
      >
        <option
          value="arcseconds"
        >
          Arcseconds
        </option>
        <option
          value="arcminutes"
        >
          Arcminutes
        </option>
        <option
          value="degrees"
        >
          Degrees
        </option>
      </SelectField>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`TargetForm render correctly 2`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Target name
      </p>
      <InputField
        className="target-name-input"
        disabled={false}
        error=""
        name="name"
        onChange={[Function]}
        value="hello"
      />
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div>
          <p>
            Resolver
          </p>
          <SelectField
            className="resolver-select"
            name="resolver"
            onChange={[Function]}
            value="Simbad"
          >
            <option
              value="Simbad"
            >
              Simbad
            </option>
            <option
              value="NED"
            >
              NED
            </option>
            <option
              value="VizieR"
            >
              VizieR
            </option>
          </SelectField>
        </styled.div>
        <styled.div>
          <br />
          <button
            className="button is-info false"
            disabled={false}
            onClick={[Function]}
            title=""
            type="button"
          >
            resolve
          </button>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Right ascension
      </p>
      <InputField
        className="right-ascension-input"
        disabled={false}
        name="rightAscension"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Declination
      </p>
      <InputField
        className="declination-input"
        disabled={false}
        name="declination"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Search radius
      </p>
      <InputField
        className="search-cone-radius-input"
        name="searchConeRadius"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Radius units
      </p>
      <SelectField
        className="radius-units-select"
        name="radiusUnits"
        onChange={[Function]}
      >
        <option
          value="arcseconds"
        >
          Arcseconds
        </option>
        <option
          value="arcminutes"
        >
          Arcminutes
        </option>
        <option
          value="degrees"
        >
          Degrees
        </option>
      </SelectField>
    </styled.div>
  </styled.div>
</Fragment>
`;

exports[`TargetForm render correctly 3`] = `
<Fragment>
  <styled.div>
    <styled.div>
      <p>
        Target name
      </p>
      <InputField
        className="target-name-input"
        disabled={false}
        error="this is an error"
        name="name"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <styled.div>
        <styled.div>
          <p>
            Resolver
          </p>
          <SelectField
            className="resolver-select"
            name="resolver"
            onChange={[Function]}
            value="Simbad"
          >
            <option
              value="Simbad"
            >
              Simbad
            </option>
            <option
              value="NED"
            >
              NED
            </option>
            <option
              value="VizieR"
            >
              VizieR
            </option>
          </SelectField>
        </styled.div>
        <styled.div>
          <br />
          <button
            className="button is-info false"
            disabled={false}
            onClick={[Function]}
            title=""
            type="button"
          >
            resolve
          </button>
        </styled.div>
      </styled.div>
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Right ascension
      </p>
      <InputField
        className="right-ascension-input"
        disabled={false}
        name="rightAscension"
        onChange={[Function]}
        value=""
      />
    </styled.div>
    <styled.div>
      <p>
        Declination
      </p>
      <InputField
        className="declination-input"
        disabled={false}
        name="declination"
        onChange={[Function]}
        value=""
      />
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <p>
        Search radius
      </p>
      <InputField
        className="search-cone-radius-input"
        name="searchConeRadius"
        onChange={[Function]}
        value="12.1"
      />
    </styled.div>
    <styled.div>
      <p>
        Radius units
      </p>
      <SelectField
        className="radius-units-select"
        name="radiusUnits"
        onChange={[Function]}
      >
        <option
          value="arcseconds"
        >
          Arcseconds
        </option>
        <option
          value="arcminutes"
        >
          Arcminutes
        </option>
        <option
          value="degrees"
        >
          Degrees
        </option>
      </SelectField>
    </styled.div>
  </styled.div>
</Fragment>
`;
